### ISSUE TITLE
tests: improve frontend tests

### ISSUE THREAD

**Issue by eamon0989** [alexandria-reader/frontend#286]
The problem with our frontend tests is that the depend on the backend database being in a certain state. Currently, the tests do not pass and are skipped. Instead, we should mock the api responses and ensure everything is correctly rendered. 

One example of how to do so is shown here: https://www.digitalocean.com/community/tutorials/how-to-test-a-react-app-with-jest-and-react-testing-library

**Comment by mrchrmn** [comment]
I agree. Testing is the area in where I had to level up and learn the most in my first couple of months. Especially how to mock dependencies away for pure unit tests,

**Comment by eamon0989** [comment]
I don’t work with React at work, how do you generally go about doing testing? Do you test the components or the services? We also mock dependencies on the back end so that part shouldn’t be too hard.

**Comment by medic-code** [comment]
Take everything below as something you're far more capable than I, having a lot more experience, but thought I'd give you some of my own experience doing this with my own App. 

Definitely sounds like you want to mock the api responses as it sounds like the tests are a little brittle. It's also worth considering some CI to only be able to push when specific tests/all tests pass. Its also possible to use husky as a pre-commit to only commit when tests pass (but this can slow developer time down).

There's a couple of options for mocking

1. Mocking with Jest - you can mock an entire module like axios, mock specific parts like get method
2. Nock
3. Mock Service Workers
4. Use a test database instead of prod - setup the database prior to tests running (delete previous records and input select data into database)

Nock and MSW are mocking libraries they both work by monkey patching http requests, although MSW does intercept real network requests in the browser. Nock intercepts node's http request function and allows you to specify what gets returned.

Using a test database is truer to a real request and so captures problems on a real database. The problem being is that the testing will be slower. 

Personal Experience 
I've used nock mostly because it's slightly easier to set up, doesn't require loading up a server you can just install and go. 
Nock is one step removed from the source of the backend compared to Mock Service Workers. MSW is as close to the backend as you can to mimicking network requests. Personally using Nock + Testing library has been what I've used its been relatively painless for unit and integration type tests. Cypress for e2e.

It would be worth having a discussion about what you would want to test, are you unit testing heavy + e2e/integration testing heavy ? Whats your opinion on testing props/handlers being called. Are there things you don't want to tests ? What are they ? 

Tests both components and features (either by e2e or an "integration" type test). At a component level it can be quite easy to test implementation details.
